maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -388,3 +429,723 @@ def clear_cache(self):
+        # serialization_fn=lambda x: x.serialize(),
+        loading_fn=lambda x: _load_tokenizer_element(x, PathTokenizers),
+    )
+
+    def __repr__(self) -> str:
+        return "-".join(
+            [type(self).__name__, *(repr(el) for el in self._tokenizer_elements)]
+        )
+
+    @cached_property
+    def _tokenizer_elements(self):
+        return [
+            self.prompt_sequencer,
+            self.coord_tokenizer,
+            self.adj_list_tokenizer,
+            self.target_tokenizer,
+            self.path_tokenizer,
+        ]
+
+    @property
+    def name(self) -> str:
+        """Serializes MazeTokenizer into a key for encoding in zanj"""
+        return repr(self)
+
+    def summary(self) -> dict[str, Any]:
+        return {
+            name: getattr(self, name)
+            for name in [
+                "prompt_sequencer",
+                "coord_tokenizer",
+                "adj_list_tokenizer",
+                "target_tokenizer",
+                "path_tokenizer",
+            ]
+        }
+
+    @classmethod
+    def from_name(cls, key: str) -> "MazeTokenizer2":
+        """Builds a MazeTokenizer from the output of `MazeTokenizer2.name`"""
+        raise NotImplementedError
+
+    @classmethod
+    def from_legacy(
+        cls, legacy_maze_tokenizer: MazeTokenizer | TokenizationMode
+    ) -> "MazeTokenizer2":
+        if isinstance(legacy_maze_tokenizer, MazeTokenizer):
+            legacy_maze_tokenizer = legacy_maze_tokenizer.tokenization_mode
+        return {
+            TokenizationMode.AOTP_UT_uniform: MazeTokenizer2(),
+            TokenizationMode.AOTP_UT_rasterized: MazeTokenizer2(),
+            TokenizationMode.AOTP_CTT_indexed: MazeTokenizer2(
+                coord_tokenizer=CoordTokenizers.CTT()
+            ),
+        }[legacy_maze_tokenizer]
+
+    def to_tokens(
+        self,
+        conn_list: ConnectionList,
+        origin: Coord | None,
+        target: Iterable[Coord],
+        path: CoordArray | None,
+    ) -> list[str]:
+        """Converts a set of maze elements into a list of tokens.
+        """
+        return self.prompt_sequencer.to_tokens(
+            conn_list,
+            origin,
+            target,
+            path,
+            self.coord_tokenizer,
+            self.adj_list_tokenizer,
+            self.target_tokenizer,
+            self.path_tokenizer,
+        )
+
+    @classmethod
+    def from_tokens(
+        cls,
+        tokens: str | list[str],
+    ) -> "MazeTokenizer2":
+        """
+        Infers most MazeTokenizer parameters from a full set of tokens.
+        Could be useful for adapting old code to new `MazeTokenizer`.
+        Would probably need a couple of other pieces of info besides just tokens.
+        - max_grid_size
+        - rasterization_mode: Only needed if UT tokens
+        - Anything else?
+        """
+        # Don't need directly, but something similar needed for LatticeMaze.from_tokens
+        raise NotImplementedError
```

Comment:
shouldn't this be something like `raise NotImplementedError("recovering maze objects from MazeTokenizer2-produces strings not implemented yet")`